### PR TITLE
BABL -> JSON schema

### PR DESCRIPTION
**Pull request by fterrazzoni**
Export a JSON-Schema from a BABL template. See http://json-schema.org/ for more details. It also comes with a few minor improvements.

External changes :
- Added `CompiledTemplate#json_schema` returns a JSON schema validating all possible outputs of the BABL template. The old `CompiledTemplate#documentation` has been removed.

- Stricter BABL template validation: `#merge` will fail at compilation time if BABL determines that there exists a case where one of the arguments cannot be merged. For instance:
```ruby
merge(
   switch(test? => { value: 42 }, default => 'not a hash'), 
   object(devil: 666)
)
```
This template is now invalid, because if `test?` returns false, `#merge` will be called on a string. One could argue that it might never happen, but in this case, the suspicious string will never be used anyway... which was probably not intended.

- `extends(partial, added_properties...)` now accepts any partial, even if it is not producing a JSON object, as long as there is no `added_properties` of course.

- Added contextual information to the errors `Only primitives can be serialized` and `Not enumerable`.

Internal changes : 
- Better separation between `Nodes`, `Operators` and `Schema`.
- Added a bunch of new test cases covering more edge cases.
- Depends on [Values](https://github.com/tcrayford/Values) gem. 

Next steps : 
- Implement type annotations to produce more useful schemas.

### CODE REVIEW

File: lib/babl/builder/template_base.rb [hunk 1]
```diff
@@ -9,8 +14,16 @@ def initialize(builder = ChainBuilder.new(&:itself))
                 @builder = builder
             end
 
-            def compile(**options)
-                Rendering::CompiledTemplate.new(precompile, **options)
+            def compile(preloader: Rendering::NoopPreloader, pretty: true)
```
Comment (vmeyet) [inline lib/babl/builder/template_base.rb]: does this mean that for bannerman we should have pretty false?

imo false would be a better default

File: lib/babl/schema/any_of.rb [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+            end
+
+            def json
+                { anyOf: choices.map(&:json) }
+            end
+
+            # Perform simple transformations in order to reduce the size of the generated
+            # schema.
+            def simplify
+                simplify_single ||
```
Comment (vmeyet) [inline lib/babl/schema/any_of.rb]: so if the simplification is not available we don't try the other simplification

File: lib/babl/builder/template_base.rb (hunk 1 above)
Comment (fterrazzoni) [inline lib/babl/builder/template_base.rb]: We are already doing :
```ruby
::Babl.configure do |config|
    config.pretty = Rails.env.development?
end
```

File: lib/babl/schema/any_of.rb (hunk 2 above)
Comment (fterrazzoni) [inline lib/babl/schema/any_of.rb]: Each simplification method is responsible for returning the simplest schema, and they do that by recursively calling `simplify` on the result after applying their own transformations, if the result is of type `AnyOf`.
